### PR TITLE
Change the way QE alloc/free resource group slot.

### PR DESCRIPTION
**Pull request by guofengrichard**
Previously QD dispatches resource group slot id to QEs
and each QE gets a slot according to the slot id.

The problem with this way is if QD exits before QE and
then dispatches the same slot id in a new session, two
different sessions on QE may share the same slot.

In this commit, QD no longer dispatches slot id, each QE
alloc/free resource group slot from its own slot pool.

Signed-off-by: xiong-gang <gxiong@pivotal.io>

### CODE REVIEW

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -301,6 +279,18 @@ static void resgroupDumpCaps(StringInfo str, ResGroupCap *caps);
...
+static ResGroupSlotData *sessionGetSlot(void);
+static void sessionResetSlot(void);
+
+#ifdef USE_ASSERT_CHECKING
+static bool selfIsAssigned(void);
+static bool selfHasSlot(void);
+static void slotValidate(const ResGroupSlotData *slot);
+static bool slotIsInUse(const ResGroupSlotData *slot);
+static bool groupIsNotDropped(const ResGroupData *group);
+#endif//USE_ASSERT_CHECKING
```
Comment (paul-guo-) [inline src/backend/utils/resgroup/resgroup.c]: We better not use c++ // for comment although gcc supports it for C.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -3432,10 +3279,38 @@ resgroupDumpFreeSlots(StringInfo str)
...
 	appendStringInfo(str, "\"free_slot_root\":{");
-	appendStringInfo(str, "\"prev\":%d,", slotGetId(root->prev));
 	appendStringInfo(str, "\"next\":%d", slotGetId(root->next));
 	appendStringInfo(str, "}");
 }
+
+/*
+ * Set resource group slot for current session.
+ */
+static void sessionSetSlot(ResGroupSlotData *slot)
```
Comment (paul-guo-) [inline src/backend/utils/resgroup/resgroup.c]: I noticed that the return type is sometimes on the same line of function declaration and sometimes isn't in this file. Better unify.

File: src/backend/utils/resgroup/resgroup.c (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: I don't see any sense to group these assert-only functions, as some of them might be used in non-assert cases during the development, so we have to move them out again?

Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: Make sense. Will change that.

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: Good catch. The return type should be in a different line here.

File: src/backend/utils/resgroup/resgroup.c [hunk 3]
```diff
@@ -2704,9 +2614,8 @@ ResGroupCapsToOpts(const ResGroupCaps *capsIn, ResGroupOpts *optsOut)
 /*
  * Validate the consistency of the resgroup information in self.
  *
- * This function checks the consistency of (slot & slotId) and
- * (group & groupId), but it does not require slot and group to be
- * both set or unset.
+ * This function checks the consistency of (group & groupId),
+ * but it does not require slot and group to be both set or unset.
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: this `but it...` desc can be retired as slot is not mentioned at all in this function.

File: src/backend/utils/resgroup/resgroup.c [hunk 4]
```diff
@@ -2836,17 +2743,14 @@ selfUnassignDroppedGroup(void)
 /*
  * Check whether self has been set a slot.
  *
- * Consistency will be checked on the slotId and slot pointer.
+ * Consistency will be checked on the slot pointer.
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: retire this desc as there is no check at all.

File: src/backend/utils/resgroup/resgroup.c [hunk 5]
```diff
@@ -2988,53 +2887,33 @@ procWakeup(PGPROC *proc)
...
-
 		Assert(slot->sessionId == InvalidSessionId);
 		Assert(slot->groupId == InvalidOid);
 		Assert(slot->nProcs == 0);
 		Assert(slot->memQuota < 0);
 		Assert(slot->memUsage == 0);
 	}
+	else
+	{
+		Assert(slot != pResGroupControl->freeSlot);
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: I guess the purpose is to check whether `slot` is in the free list, but above check does not implement this at all. When slot pool was a double linked queue this `freeSlot` was used as the root which is never a valid node, but as now we have changed it to a stack again I don't see any point to check a slot with it.

File: src/backend/utils/resgroup/resgroup.c [hunk 6]
```diff
@@ -3074,27 +2935,17 @@ slotGetId(const ResGroupSlotData *slot)
 {
 	int			slotId;
 
-	if (slot == NULL || slot == &pResGroupControl->freeSlot)
+	if (slot == NULL || slot == pResGroupControl->freeSlot)
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: similar as the comment for `slotValidate()`

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: `static void` should be put on a separate line.

File: src/include/utils/session_state.h [hunk 7]
```diff
@@ -88,6 +88,11 @@ typedef struct SessionState
 	 * next entry would point to the next used entry */
 	struct SessionState *next;
 
+	/*
+	 * Resource group per-session slot information.
+	 */
+	struct ResGroupSlotData *resGroupSlot;
```
Comment (pivotal-ning-yu) [inline src/include/utils/session_state.h]: Where is `ResGroupSlotData` defined?

File: src/backend/utils/resgroup/resgroup.c (hunk 3 above)
Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: Makes sense. Will do it.

File: src/backend/utils/resgroup/resgroup.c (hunk 4 above)
Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: Makes sense. Will do it.

File: src/backend/utils/resgroup/resgroup.c (hunk 5 above)
Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: If a slot is in use, it should not be in the free list. Specifically, it should not be equal to the head of the free list. 
This assertion may catch some mistakes. I prefer to keep it.

File: src/include/utils/session_state.h (hunk 7 above)
Comment (guofengrichard) [reply src/include/utils/session_state.h]: Good catch. Will use a `void *` here.

File: src/backend/utils/resgroup/resgroup.c (hunk 5 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: -1

Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: If you do want to check sth. I would suggest to make a real check by walking through the free list, you can put this check in a `#ifdef` so it doesn't affect the release version. And you could also check for slot->next, at least better than current one.

**Comment by pivotal-ning-yu** [comment]
LGTM

File: src/backend/utils/resgroup/resgroup.c [hunk 8]
```diff
@@ -2247,6 +2242,9 @@ AssignResGroupOnMaster(void)
 
 		Assert(!self->doMemCheck);
 
+		/* Set resource group slot for current session */
+		sessionSetSlot(slot);
```
Comment (pengzhout) [inline src/backend/utils/resgroup/resgroup.c]: A test that for a session has been bumped should be good here.

Comment (guofengrichard) [reply src/backend/utils/resgroup/resgroup.c]: I tested several cases of session bump. After `CheckForResetSession()` is called, `UnassignResGroup()` will then be called and the slot of current session will be returned to slot pool.

**Pull request merged**